src/USER-MISC/fix_charge_regulation.cpp

```diff
@@ -0,0 +1,1335 @@
+    if (exclusion_group == -1)
+      error->all(FLERR, "Could not find fix CR exclusion group ID");
+    exclusion_group_bit = group->bitmask[exclusion_group];
+
+    // neighbor list exclusion setup
+    // turn off interactions between group all and the exclusion group
+
+    int narg = 4;
+    char **arg = new char *[narg];;
+    arg[0] = (char *) "exclude";
+    arg[1] = (char *) "group";
+    arg[2] = group_arg[0];
+    arg[3] = (char *) "all";
+    neighbor->modify_params(narg, arg);
+    delete[] group_arg[0];
+    delete[] group_arg;
+    delete[] arg;
+  }
+
+  // check that no deletable atoms are in atom->firstgroup
+  // deleting such an atom would not leave firstgroup atoms first
+
+  if (atom->firstgroup >= 0) {
+    int *mask = atom->mask;
+    int firstgroupbit = group->bitmask[atom->firstgroup];
+
+    int flag = 0;
+    for (int i = 0; i < atom->nlocal; i++)
+      if ((mask[i] == groupbit) && (mask[i] && firstgroupbit)) flag = 1;
+
+    int flagall;
+    MPI_Allreduce(&flag, &flagall, 1, MPI_INT, MPI_SUM, world);
+
+    if (flagall)
+      error->all(FLERR, "Cannot do Fix charge regulation on atoms in atom_modify first group");
+  }
+
+  // construct group bitmask for all new atoms
+  // aggregated over all group keywords
+
+  groupbitall = 1 | groupbit;
+
+  for (int igroup = 0; igroup < ngroups; igroup++) {
+    int jgroup = group->find(groupstrings[igroup]);
+    if (jgroup == -1)
+      error->all(FLERR, "Could not find specified fix charge regulation group ID");
+    groupbitall |= group->bitmask[jgroup];
+  }
+}
+
+void Fix_charge_regulation::pre_exchange() {
+
+  if (next_reneighbor != update->ntimestep) return;
+  xlo = domain->boxlo[0];
+  xhi = domain->boxhi[0];
+  ylo = domain->boxlo[1];
+  yhi = domain->boxhi[1];
+  zlo = domain->boxlo[2];
+  zhi = domain->boxhi[2];
+
+  if (triclinic) {
+    sublo = domain->sublo_lamda;
+    subhi = domain->subhi_lamda;
+  } else {
+    sublo = domain->sublo;
+    subhi = domain->subhi;
+  }
+  volume = domain->xprd * domain->yprd * domain->zprd;
+  if (triclinic) domain->x2lamda(atom->nlocal);
+  domain->pbc();
+  comm->exchange();
+  atom->nghost = 0;
+  comm->borders();
+
+  if (triclinic) domain->lamda2x(atom->nlocal + atom->nghost);
+  energy_stored = energy_full();
+  if (energy_stored > MAXENERGYTEST)
+    error->warning(FLERR, "Energy of old configuration in fix charge_regulation is > MAXENERGYTEST.");
+
+  if ((reaction_distance > fabs(domain->boxhi[0] - domain->boxlo[0]) / 2) ||
+      (reaction_distance > fabs(domain->boxhi[1] - domain->boxlo[1]) / 2) ||
+      (reaction_distance > fabs(domain->boxhi[2] - domain->boxlo[2]) / 2)) {
+    error->warning(FLERR,
+                   "reaction distance (rxd) is larger than half the box dimension, resetting default: xrd = 0.");
+    reaction_distance = 0;
+  }
+  // volume in units of (N_A * mol / liter)
+  volume_rx = (xhi - xlo) * (yhi - ylo) * (zhi - zlo) * pow(llength_unit_in_nm, 3) * 0.602214;
```

Comment:
What does this constant 0.602214 represent? I have an educated guess, but this should be using a symbolic name defined at the top of the file so that it is explicit and obvious to the reader.